examples/2-clicks/README.md

```diff
@@ -0,0 +1,203 @@
+
+```go
+func runView() {
+	view, err := goka.NewView(brokers,
+		goka.GroupTable(group),
+		new(userCodec),
+	)
+	if err != nil {
+		panic(err)
+	}
+	go view.Start()
+	defer view.Stop()
+
+	root := mux.NewRouter()
+	root.HandleFunc("/{key}", func(w http.ResponseWriter, r *http.Request) {
+		value, _ := view.Get(mux.Vars(r)["key"])
+		data, _ := json.Marshal(value)
+		w.Write(data)
+	})
+	fmt.Println("View opened at http://localhost:9095/")
+	http.ListenAndServe(":9095", root)
+}
+```
+
+Finally, an `Emitter` is used to simulate user clicks. This allows us to write objects conveniently to a specific topic. The key for each message is the user ID which is simply generated by the sender loop.
+
+```go
+func runEmitter() {
+	emitter, err := goka.NewEmitter(brokers, topic,
+		new(codec.String))
+	if err != nil {
+		panic(err)
+	}
+	defer emitter.Finish()
+
+	t := time.NewTicker(100 * time.Millisecond)
+	defer t.Stop()
+
+	var i int
+	for range t.C {
+		key := fmt.Sprintf("user-%d", i%10)
+		value := fmt.Sprintf("%s", time.Now())
+		emitter.EmitSync(key, value)
+		i++
+	}
+}
+```
+
+The `Emitter` is created specifying the topic and a `Codec` that marshals passed messages for us into Kafka.
+Here we use a codec provided by goka called `codec.String`, that simply marshals from `string` values.
+In our case that's sufficient since a payload of a click is simply a time-string. If we wanted to write more
+complex objects (e.g., structs) we would have to implement our own codec, similar to the `userCodec` mentioned below.
+
+### Codec
+
+A codec is an interface that encodes and decodes between an arbitrary value and `[]byte`. This is used to store data in kafka and on disk.
+```go
+type Codec interface {
+	Encode(value interface{}) (data []byte, err error)
+	Decode(data []byte) (value interface{}, err error)
+}
+```
+Convenience Codecs for often-used types are provided by goka, like `codec.Int64` and `codec.String`. In most cases however the data is
+complex and custom codecs must be provided. In this example we implemented a userCodec that uses the `json`-Marshal/Unmarshal functionality for encoding and decoding.
+
+Note that errors returned by the codec lead to a shutdown of the
+processor/view/emitter using it immediately. We chose that fail-early-approach since data corruption would occur if,
+for example, a processor accidentally reads and writes using a wrong codec and mixes different codecs in the group table.
+If you need to tolerate codec-errors, you'll have to handle them inside the codec and make sure it returns a `nil`-error.
+
+### Partitioning and Concurrency
+
+Messages are partitioned in Kafka using the message key. Messages within each partition are processed sequentially,
+but in parallel for all partitions. That means for our case, different users can be modified
+in parallel. A single user however is always modified sequentially, because all messages for that user are assigned
+to the same partition which runs sequentially.
```

Comment:
Perhaps we should add a small explanation here as to why the messages for the user are assigned to the same partition, ie. ``` Since we are using the user id as the message's key...```